src/iterator.c

```diff
@@ -1482,6 +1482,427 @@ int git_iterator_for_workdir_ext(
+	GIT_UNUSED(entry); GIT_UNUSED(self);
+	return GIT_ENOTFOUND;
+}
+
+static int workdirfilelist_iterator__seek(
+	git_iterator *self, const char *prefix)
+{
+	GIT_UNUSED(self); GIT_UNUSED(prefix);
+	return -1;
+}
+
+static int workdirfilelist_iterator__reset(
+	git_iterator *self, const char *start, const char *end)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+
+	if (iterator__reset_range(self, start, end) < 0)
+		return -1;
+
+	/* We lookup the starting and ending paths in the filelist.
+	 * This gives us the bounds on set
+	 * of possible paths we will return during the iteration.
+	 *
+	 * For the "end" string, we want it to be included in the
+	 * set of paths enumerated (unlike the STL).  We set the
+	 * "pos_filelist_end" to be just past it (like the STL).
+	 */
+
+	wdfi->pos_filelist_start = 0;
+	if (wdfi->base.start)
+		git_vector_bsearch(&wdfi->pos_filelist_start, &wdfi->filelist, wdfi->base.start);
+
+	wdfi->pos_filelist_end = git_vector_length(&wdfi->filelist);
+	if (wdfi->base.end)
+		if (git_vector_bsearch(&wdfi->pos_filelist_end, &wdfi->filelist, wdfi->base.end) == 0)
+			wdfi->pos_filelist_end++;
+
+	/* Seed "current" with the first item in both the filelist and the workdir. */
+	/* Skip unmatched items in filelist. */
+
+	wdfi->pos_filelist_current = wdfi->pos_filelist_start;
+	while (!workdirfilelist_iterator__at_end(self)) {
+		struct stat st;
+		const char *sz = git_vector_get(&wdfi->filelist, wdfi->pos_filelist_current);
+		if (workdirfilelist_stat(&st, self, sz) == 0) {
+			git_buf_sets(&wdfi->buf_entry_path, sz);
+			wdfi->entry.path = wdfi->buf_entry_path.ptr;
+			git_index_entry__init_from_stat(&wdfi->entry, &st, true);
+			break;
+		}
+		wdfi->pos_filelist_current++;
+	}
+	return (workdirfilelist_iterator__at_end(self) ? GIT_ITEROVER : 0);
+}
+
+static void workdirfilelist_iterator__free(git_iterator *self)
+{
+	workdirfilelist_iterator *wdfi = (workdirfilelist_iterator *)self;
+	git_buf_free(&wdfi->buf_workdir);
+	git_buf_free(&wdfi->buf_entry_path);
+	git_vector_free(&wdfi->filelist);
+}
+
+int git_iterator_for_workdirfilelist(
+	git_iterator **iter,
+	git_repository *repo,
+	const char *repo_workdir,
+	const git_strarray *paths,
+	git_iterator_flag_t flags,
+	const char *start,
+	const char *end)
+{
+	int error = 0;
+	workdirfilelist_iterator *wdfi = NULL;
+
+	assert(paths && paths->count > 0);
+
+	/* This iterator only returns items from the filelist
+	 * that happen to appear in the index, so we don't do
+	 * any of these variations.
+	 */
+	assert((flags & GIT_ITERATOR_INCLUDE_TREES) == 0);
+	assert((flags & GIT_ITERATOR_DONT_AUTOEXPAND) == 0);
+	assert((flags & GIT_ITERATOR_PRECOMPOSE_UNICODE) == 0);
+
+	/* TODO We DO NOT have "tree" or "index" arguments since these
+	 * seem to be used in git_iterator_for_workdir_ext() for submodule
+	 * checking.  Our assumptiong is that the "filelist" will not
+	 * cross into submodules. Do we need to revisit this decision?
+	 */
```

Comment:
The tree and/or index are the source of truth for whether a directory is a submodule. See #2520 for what happened when we treated the `.gitmodules` file as a source of truth for that.

I think you might still need something like that here. If we have an update for a submodule in a merge or whatever, the list of paths would have this submodule, I believe. You use `git_index_entry__init_from_stat()` which would not give you the right answer in case you find a submodule, which would be in the file list but be a directory in the filesystem.